myia/infer/inferrers.py

```diff
@@ -47,52 +63,486 @@ def partial_inferrer(node, args, unif):
+    return res
+
+
+def _bin_op_inferrer(bin_op, bin_rop, node, args, unif):
+    """Generic inferrer for builtin binary operator functions."""
+    a_node, b_node = args
+    a_type = yield Require(a_node)
+    b_type = yield Require(b_node)
+
+    if isinstance(a_type, data.GenericBase) or isinstance(
+        b_type, data.GenericBase
+    ):
+        # If there is any generic in operands, just unify them.
+        return autils.unify(a_type, b_type)[0]
+
+    a_interface = a_type.tracks.interface
+    b_interface = b_type.tracks.interface
+    if hasattr(a_interface, bin_op):
+        new_node = node.graph.apply(
+            getattr(a_interface, bin_op), a_node, b_node
+        )
+    elif hasattr(b_interface, bin_rop):
+        new_node = node.graph.apply(
+            getattr(b_interface, bin_rop), b_node, a_node
+        )
+    else:
+        raise TypeError(
+            f"No {bin_op} method for {a_interface} and no {bin_rop} method for {b_interface}"
+        )
+    yield Replace(new_node)
+    res = yield Require(new_node)
+    return res
+
+
+def _bin_rop_inferrer(bin_rop, node, args, unif):
+    """Generic inferrer for binary rop methods.
+
+    Replace node with a call to rop method if available in right operand type.
+    """
+    a_node, b_node = args
+    b_type = yield Require(b_node)
+    b_interface = b_type.tracks.interface
+    if hasattr(b_interface, bin_rop):
+        new_node = node.graph.apply(
+            getattr(b_interface, bin_rop), b_node, a_node
+        )
+        yield Replace(new_node)
+        res = yield Require(new_node)
+        return res
+    else:
+        raise TypeError(f"No {bin_rop} method for {b_interface}")
+
+
+def _bin_op_dispatcher(bin_rop, *signatures):
+    """Dispatcher for binary op methods.
+
+    Create an inference function using given signatures and a
+    fallback binary rop inference function.
+    """
+
+    def inf(node, args, unif):
+        return _bin_rop_inferrer(bin_rop, node, args, unif)
+
+    return dispatch_inferences(*signatures, default=inf)
+
+
+def _bin_op_cast_right(cls, bin_op, node, args, unif):
+    """Inferrer for binary op/rop methods (e.g. float.__mul__).
+
+    Cast right operand to left operand type if necessary.
+    """
+    a_node, b_node = args
+    a_type = yield Require(a_node)
+    b_type = yield Require(b_node)
+    a_interface = a_type.tracks.interface
+    b_interface = b_type.tracks.interface
+    assert a_interface is cls, f"expected {cls}, got {a_interface}"
+    if b_interface is cls:
+        return data.AbstractAtom({"interface": cls})
+    else:
+        b_casted = node.graph.apply(cls, b_node)
+        new_node = node.graph.apply(getattr(cls, bin_op), a_node, b_casted)
+        yield Replace(new_node)
+        res = yield Require(new_node)
+        return res
+
+
+def operator_add_inferrer(node, args, unif):
+    """Inferrer for the operator.add function."""
+    return _bin_op_inferrer("__add__", "__radd__", node, args, unif)
```

Comment:
It would be more economical to have `_bin_op_inferrer("__add__", "__radd__")` return the inferrer, and write that directly in the inferrers directly. (Same for below).

Reply:
Done!